charmhelpers/contrib/openstack/deferred_events.py

```diff
@@ -0,0 +1,291 @@
+        :param action: Action that was deferred
+        :type action: str
+        """
+        filtered_restarts = [
+            e
+            for e in self.events
+            if not (e.service in services and e.action == action)]
+        self.events = filtered_restarts
+        self.save_events()
+
+    def save_events(self):
+        """Write deferred events to backend."""
+        # Use dict() to ensure the result is a dict not an OrderedDict which
+        # is returned on 3.1 <= python < 3.8
+        raw_events = [dict(e._asdict()) for e in self.events]
+        with ch_core.unitdata.HookData()() as t:
+            kv = t[0]
+            kv.set(self.KV_KEY, raw_events)
+
+
+class DeferredPackageServiceEvents(DeferredServiceEvents):
+    """Deferred service events from package actions."""
+
+    def load_events(self):
+        """Load events previously requested by the charm.
+
+        :returns: List of events requested by the charm.
+        :rtype: List[ServiceEvent]
+        """
+        queued_restarts = []
+        for event in policy_rcd.policy_deferred_events():
+            queued_restarts.append(
+                ServiceEvent(
+                    event['time'],
+                    event['service'],
+                    'Pkg Update',
+                    event['action']))
+        return queued_restarts
+
+    def clear_deferred_events(self, services, action):
+        """Clear deferred events.
+
+        :param services: Services with deferred actions to clear.
+        :type services: List[str]
+        :param action: Action that was deferred
+        :type action: str
+        """
+        policy_rcd.clear_deferred_pkg_events(services, action)
+
+
+def get_deferred_events():
+    """Return a list of deferred events requested by the charm and packages.
+
+    :returns: List of deferred events
+    :rtype: List[ServiceEvent]
+    """
+    return DeferredCharmServiceEvents().events + DeferredPackageServiceEvents().events
+
+
+def get_deferred_restarts():
+    """List of deferred restart events requested by the charm and packages.
+
+    :returns: List of deferred restarts
+    :rtype: List[ServiceEvent]
+    """
+    return [e for e in get_deferred_events() if e.action == 'restart']
+
+
+def clear_deferred_events(services, action):
+    """Clear deferred events of type `action` targetted at `services`.
+
+    :param services: Services with deferred actions to clear.
+    :type services: List[str]
+    :param action: Action to clear
+    :type action: str
+    """
+    DeferredCharmServiceEvents().clear_deferred_events(services, action)
+    DeferredPackageServiceEvents().clear_deferred_events(services, action)
+
+
+def clear_deferred_restarts(services):
+    """Clear deferred restart events targetted at `services`.
+
+    :param services: Services with deferred actions to clear.
+    :type services: List[str]
+    """
+    clear_deferred_events(services, 'restart')
+
+
+def process_svc_restart(service):
```

Comment:
I am a little fuzzy why not call clear_deferred_restarts directly? Is this future-proofing for more actions to be taken? Or maybe it will become clearer to me as I read more.

Reply:
Yes, future proofing was my thinking. The restart_on_change decorator calls this function after a restart event. I figured we might want to do more than just clear the event